Crash after upgrade to v5.1.0: Panic due to nil pointer dereference

### Describe the bug

After upgrading to version `5.1.0`, the plugin crashes with a panic caused by a nil pointer dereference. The issue appears to occur during the usage of `setRealmData`, within the `Compare` and `GreaterThanOrEqual` methods when handling a `nil` Keycloak version object. I initially observed the crash when the plugin attempted to refresh the `keycloak_realm` resource (see stack trace below), however, I tried to temporarily disable the realm resource and started getting the same crash on other resources, so there is probably another underlying issue. Happy to provide more details if needed.

```
Stack trace from the terraform-provider-keycloak_v5.1.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference

[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x88f613]

goroutine 42 [running]:

[github.com/hashicorp/go-version.(*Version](http://github.com/hashicorp/go-version.(*Version)).String(0x0)

	[github.com/hashicorp/go-version@v1.7.0/version.go:386](http://github.com/hashicorp/go-version@v1.7.0/version.go:386) +0x33

[github.com/hashicorp/go-version.(*Version](http://github.com/hashicorp/go-version.(*Version)).Compare(0x0, 0xc0005a0b90)

	[github.com/hashicorp/go-version@v1.7.0/version.go:119](http://github.com/hashicorp/go-version@v1.7.0/version.go:119) +0x25

[github.com/hashicorp/go-version.(*Version](http://github.com/hashicorp/go-version.(*Version)).GreaterThanOrEqual(...)

	[github.com/hashicorp/go-version@v1.7.0/version.go:320](http://github.com/hashicorp/go-version@v1.7.0/version.go:320)

[github.com/keycloak/terraform-provider-keycloak/provider.setRealmData](http://github.com/keycloak/terraform-provider-keycloak/provider.setRealmData)(0xc000198e80, 0xc0001b9408, 0x0)

	[github.com/keycloak/terraform-provider-keycloak/provider/resource_keycloak_realm.go:1314](http://github.com/keycloak/terraform-provider-keycloak/provider/resource_keycloak_realm.go:1314) +0x2cc6

[github.com/keycloak/terraform-provider-keycloak/provider.resourceKeycloakRealmRead](http://github.com/keycloak/terraform-provider-keycloak/provider.resourceKeycloakRealmRead)({0xfa8310, 0xc0002e62a0}, 0xc000198e80, {0xe40b40?, 0xc0004342a0})

	[github.com/keycloak/terraform-provider-keycloak/provider/resource_keycloak_realm.go:1438](http://github.com/keycloak/terraform-provider-keycloak/provider/resource_keycloak_realm.go:1438) +0x179

[github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource](http://github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource)).read(0xc00039f340, {0xfa8268, 0xc0005a9800}, 0xc000198e80, {0xe40b40, 0xc0004342a0})

	[github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/resource.go:823](http://github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/resource.go:823) +0x119

[github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource](http://github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource)).RefreshWithoutUpgrade(0xc00039f340, {0xfa8268, 0xc0005a9800}, 0xc0004f2ea0, {0xe40b40, 0xc0004342a0})

	[github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/resource.go:1117](http://github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/resource.go:1117) +0x529

[github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer](http://github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer)).ReadResource(0xc000386ba0, {0xfa8268?, 0xc0005a9740?}, 0xc0000c3800)

	[github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/grpc_provider.go:710](http://github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/grpc_provider.go:710) +0x6c5

[github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server](http://github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server)).ReadResource(0xc0002e2500, {0xfa8268?, 0xc0005a8f60?}, 0xc0002e78f0)

	[github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov5/tf5server/server.go:783](http://github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov5/tf5server/server.go:783) +0x309

[github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler](http://github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler)({0xe218c0, 0xc0002e2500}, {0xfa8268, 0xc0005a8f60}, 0xc000199900, 0x0)

	[github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:575](http://github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:575) +0x1a9

[google.golang.org/grpc.(*Server](http://google.golang.org/grpc.(*Server)).processUnaryRPC(0xc000194e00, {0xfa8268, 0xc0005a8ed0}, {0xfacdd0, 0xc0003f1040}, 0xc0005aa6c0, 0xc000410960, 0x15a7228, 0x0)

	[google.golang.org/grpc@v1.67.1/server.go:1394](http://google.golang.org/grpc@v1.67.1/server.go:1394) +0xe49

[google.golang.org/grpc.(*Server](http://google.golang.org/grpc.(*Server)).handleStream(0xc000194e00, {0xfacdd0, 0xc0003f1040}, 0xc0005aa6c0)

	[google.golang.org/grpc@v1.67.1/server.go:1805](http://google.golang.org/grpc@v1.67.1/server.go:1805) +0xe8b

[google.golang.org/grpc.(*Server](http://google.golang.org/grpc.(*Server)).serveStreams.func2.1()

	[google.golang.org/grpc@v1.67.1/server.go:1029](http://google.golang.org/grpc@v1.67.1/server.go:1029) +0x8b

created by [google.golang.org/grpc.(*Server](http://google.golang.org/grpc.(*Server)).serveStreams.func2 in goroutine 28

	[google.golang.org/grpc@v1.67.1/server.go:1040](http://google.golang.org/grpc@v1.67.1/server.go:1040) +0x125

Error: The terraform-provider-keycloak_v5.1.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely

helpful if you could report the crash with the plugin's maintainers so that it

can be fixed. The output above should help diagnose the issue.

2025-01-24T11:58:58.386Z [DEBUG] provider: plugin exited
```

### Version

26.1.0

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Anything else?

_No response_

So, I've tested with the following KC version: 26.0.5, 26.0.8, 26.1.0, 26.1.1.

Applying (for the first time) this terraform code crash:

```terraform
terraform {
  required_providers {
    keycloak = {
      source  = "keycloak/keycloak"
      version = "= 5.1.0"
    }
  }
  required_version = ">= 1.6"
}

variable "keycloak_admin_password" {
  type = string
}

locals {
  keycloak_url = "https://my-url"
  realm_name   = "provider-crash-realm-test"
}

provider "keycloak" {
  client_id                = "admin-cli"
  username                 = "admin"
  password                 = var.keycloak_admin_password
  url                      = local.keycloak_url
  initial_login            = false
}

resource "keycloak_realm" "platform_realm" {
  realm             = local.realm_name
  enabled           = true
}
```

with the following stack trace:

```
keycloak_realm.platform_realm: Creating...
╷
│ Error: Plugin did not respond
│
│   with keycloak_realm.platform_realm,
│   on keycloak.tf line 28, in resource "keycloak_realm" "platform_realm":
│   28: resource "keycloak_realm" "platform_realm" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-keycloak_v5.1.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x88f613]

goroutine 41 [running]:
github.com/hashicorp/go-version.(*Version).String(0x0)
        github.com/hashicorp/go-version@v1.7.0/version.go:386 +0x33
github.com/hashicorp/go-version.(*Version).Compare(0x0, 0xc0005e8c80)
        github.com/hashicorp/go-version@v1.7.0/version.go:119 +0x25
github.com/hashicorp/go-version.(*Version).GreaterThanOrEqual(...)
        github.com/hashicorp/go-version@v1.7.0/version.go:320
github.com/keycloak/terraform-provider-keycloak/provider.setRealmFlowBindings(0xc00062ac00, 0xc000696a08, 0x0)
        github.com/keycloak/terraform-provider-keycloak/provider/resource_keycloak_realm.go:731 +0x5d5
github.com/keycloak/terraform-provider-keycloak/provider.getRealmFromData(0xc00062ac00, 0x0)
        github.com/keycloak/terraform-provider-keycloak/provider/resource_keycloak_realm.go:1030 +0x1e71
github.com/keycloak/terraform-provider-keycloak/provider.resourceKeycloakRealmCreate({0xfa8310, 0xc000128690}, 0xc00062ac00, {0xe40b40, 0xc000114c00})
        github.com/keycloak/terraform-provider-keycloak/provider/resource_keycloak_realm.go:1399 +0x6d
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xc0003af340, {0xfa8268, 0xc000472c90}, 0xc00062ac00, {0xe40b40, 0xc000114c00})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/resource.go:806 +0x119
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0003af340, {0xfa8268, 0xc000472c90}, 0xc000550dd0, 0xc00062a900, {0xe40b40, 0xc000114c00})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/resource.go:937 +0xa89
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000396bb8, {0xfa8268?, 0xc000472bd0?}, 0xc000476370)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/grpc_provider.go:1155 +0xd5c
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0002f4500, {0xfa8268?, 0xc0004721e0?}, 0xc000128070)
        github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov5/tf5server/server.go:865 +0x3d0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0xe218c0, 0xc0002f4500}, {0xfa8268, 0xc0004721e0}, 0xc00062a080, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:611 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001b2e00, {0xfa8268, 0xc000472150}, {0xfacdd0, 0xc0004011e0}, 0xc00047c000, 0xc000422960, 0x15a7258, 0x0)
        google.golang.org/grpc@v1.67.1/server.go:1394 +0xe49
google.golang.org/grpc.(*Server).handleStream(0xc0001b2e00, {0xfacdd0, 0xc0004011e0}, 0xc00047c000)
        google.golang.org/grpc@v1.67.1/server.go:1805 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.67.1/server.go:1029 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 16
        google.golang.org/grpc@v1.67.1/server.go:1040 +0x125

Error: The terraform-provider-keycloak_v5.1.0 plugin crashed!
```

For information, I've installed the 3 following plugins in Keycloak:
- [keycloak-home-idp-discovery](https://github.com/sventorben/keycloak-home-idp-discovery)
- [keycloak-restrict-client-auth](https://github.com/sventorben/keycloak-restrict-client-auth)
- [keycloak-metrics-spi](https://github.com/aerogear/keycloak-metrics-spi)

I'll see if I can try without those plugins

--- a/keycloak/version.go
+++ b/keycloak/version.go
@@ -39,54 +39,36 @@ func (v Version) AsVersion() *version.Version {
 	return vv
 }
 
-func (KeycloakClient *KeycloakClient) Version() *version.Version {
-	return KeycloakClient.version
+func (KeycloakClient *KeycloakClient) Version(ctx context.Context) (*version.Version, error) {
+	if KeycloakClient.version == nil {
+		err := KeycloakClient.login(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return KeycloakClient.version, nil
 }
 
 func (keycloakClient *KeycloakClient) VersionIsGreaterThanOrEqualTo(ctx context.Context, versionString Version) (bool, error) {
-	if keycloakClient.version == nil {
-		err := keycloakClient.login(ctx)
-		if err != nil {
-			return false, err
-		}
-	}
-
-	v, err := version.NewVersion(string(versionString))
+	version, err := keycloakClient.Version(ctx)
 	if err != nil {
-		return false, nil
+		return false, err
 	}
-
-	return keycloakClient.version.GreaterThanOrEqual(v), nil
+	return version.GreaterThanOrEqual(versionString.AsVersion()), nil
 }
 
 func (keycloakClient *KeycloakClient) VersionIsLessThanOrEqualTo(ctx context.Context, versionString Version) (bool, error) {
-	if keycloakClient.version == nil {
-		err := keycloakClient.login(ctx)
-		if err != nil {
-			return false, err
-		}
-	}
-
-	v, err := version.NewVersion(string(versionString))
+	version, err := keycloakClient.Version(ctx)
 	if err != nil {
-		return false, nil
+		return false, err
 	}
-
-	return keycloakClient.version.LessThanOrEqual(v), nil
+	return version.LessThanOrEqual(versionString.AsVersion()), nil
 }
 
 func (keycloakClient *KeycloakClient) VersionIsLessThan(ctx context.Context, versionString Version) (bool, error) {
-	if keycloakClient.version == nil {
-		err := keycloakClient.login(ctx)
-		if err != nil {
-			return false, err
-		}
-	}
-
-	v, err := version.NewVersion(string(versionString))
+	version, err := keycloakClient.Version(ctx)
 	if err != nil {
-		return false, nil
+		return false, err
 	}
-
-	return keycloakClient.version.LessThan(v), nil
+	return version.LessThan(versionString.AsVersion()), nil
 }

--- a/provider/data_source_keycloak_realm.go
+++ b/provider/data_source_keycloak_realm.go
@@ -541,7 +541,10 @@ func dataSourceKeycloakRealm() *schema.Resource {
 
 func dataSourceKeycloakRealmRead(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
-	keycloakVersion := keycloakClient.Version()
+	keycloakVersion, err := keycloakClient.Version(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	realmName := data.Get("realm").(string)
 

--- a/provider/generic_keycloak_identity_provider.go
+++ b/provider/generic_keycloak_identity_provider.go
@@ -205,7 +205,11 @@ func resourceKeycloakIdentityProviderImport(_ context.Context, d *schema.Resourc
 func resourceKeycloakIdentityProviderCreate(getIdentityProviderFromData identityProviderDataGetterFunc, setDataFromIdentityProvider identityProviderDataSetterFunc) func(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	return func(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 		keycloakClient := meta.(*keycloak.KeycloakClient)
-		keycloakVersion := keycloakClient.Version()
+		keycloakVersion, err := keycloakClient.Version(ctx)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
 		identityProvider, err := getIdentityProviderFromData(data, keycloakVersion)
 		if err != nil {
 			return diag.FromErr(err)
@@ -224,7 +228,10 @@ func resourceKeycloakIdentityProviderCreate(getIdentityProviderFromData identity
 func resourceKeycloakIdentityProviderRead(setDataFromIdentityProvider identityProviderDataSetterFunc) func(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	return func(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 		keycloakClient := meta.(*keycloak.KeycloakClient)
-		keycloakVersion := keycloakClient.Version()
+		keycloakVersion, err := keycloakClient.Version(ctx)
+		if err != nil {
+			return diag.FromErr(err)
+		}
 		realm := data.Get("realm").(string)
 		alias := data.Get("alias").(string)
 		identityProvider, err := keycloakClient.GetIdentityProvider(ctx, realm, alias)
@@ -239,7 +246,10 @@ func resourceKeycloakIdentityProviderRead(setDataFromIdentityProvider identityPr
 func resourceKeycloakIdentityProviderUpdate(getIdentityProviderFromData identityProviderDataGetterFunc, setDataFromIdentityProvider identityProviderDataSetterFunc) func(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	return func(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 		keycloakClient := meta.(*keycloak.KeycloakClient)
-		keycloakVersion := keycloakClient.Version()
+		keycloakVersion, err := keycloakClient.Version(ctx)
+		if err != nil {
+			return diag.FromErr(err)
+		}
 		identityProvider, err := getIdentityProviderFromData(data, keycloakVersion)
 		if err != nil {
 			return diag.FromErr(err)

--- a/provider/resource_keycloak_authentication_bindings.go
+++ b/provider/resource_keycloak_authentication_bindings.go
@@ -93,7 +93,10 @@ func resetAuthenticationBindingsForRealm(realm *keycloak.Realm, keycloakVersion 
 
 func resourceKeycloakAuthenticationBindingsCreate(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
-	keycloakVersion := keycloakClient.Version()
+	keycloakVersion, err := keycloakClient.Version(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	realm, err := keycloakClient.GetRealm(ctx, data.Get("realm_id").(string))
 	if err != nil {
@@ -124,7 +127,10 @@ func resourceKeycloakAuthenticationBindingsCreate(ctx context.Context, data *sch
 
 func resourceKeycloakAuthenticationBindingsRead(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
-	keycloakVersion := keycloakClient.Version()
+	keycloakVersion, err := keycloakClient.Version(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	realm, err := keycloakClient.GetRealm(ctx, data.Id())
 	if err != nil {
@@ -138,7 +144,10 @@ func resourceKeycloakAuthenticationBindingsRead(ctx context.Context, data *schem
 
 func resourceKeycloakAuthenticationBindingsDelete(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
-	keycloakVersion := keycloakClient.Version()
+	keycloakVersion, err := keycloakClient.Version(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	realm, err := keycloakClient.GetRealm(ctx, data.Id())
 	if err != nil {
@@ -157,7 +166,10 @@ func resourceKeycloakAuthenticationBindingsDelete(ctx context.Context, data *sch
 
 func resourceKeycloakAuthenticationBindingsUpdate(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
-	keycloakVersion := keycloakClient.Version()
+	keycloakVersion, err := keycloakClient.Version(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	realm, err := keycloakClient.GetRealm(ctx, data.Id())
 	if err != nil {

--- a/provider/resource_keycloak_realm.go
+++ b/provider/resource_keycloak_realm.go
@@ -1394,7 +1394,10 @@ func getHeaderSettings(realm *keycloak.Realm) map[string]interface{} {
 
 func resourceKeycloakRealmCreate(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
-	keycloakVersion := keycloakClient.Version()
+	keycloakVersion, err := keycloakClient.Version(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	realm, err := getRealmFromData(data, keycloakVersion)
 	if err != nil {
@@ -1423,7 +1426,10 @@ func resourceKeycloakRealmCreate(ctx context.Context, data *schema.ResourceData,
 
 func resourceKeycloakRealmRead(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
-	keycloakVersion := keycloakClient.Version()
+	keycloakVersion, err := keycloakClient.Version(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	realm, err := keycloakClient.GetRealm(ctx, data.Id())
 	if err != nil {
@@ -1442,7 +1448,10 @@ func resourceKeycloakRealmRead(ctx context.Context, data *schema.ResourceData, m
 
 func resourceKeycloakRealmUpdate(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
-	keycloakVersion := keycloakClient.Version()
+	keycloakVersion, err := keycloakClient.Version(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	realm, err := getRealmFromData(data, keycloakVersion)
 	if err != nil {

